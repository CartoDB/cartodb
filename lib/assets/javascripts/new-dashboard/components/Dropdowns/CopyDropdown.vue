<template>
    <div class="dropdown" v-if="isVisible" @click.prevent="copyName">
      <span class="text is-small copy-text" v-if="!copySuccessful"><img svg-inline class="copy-icon" src="../../assets/icons/common/copy.svg">{{$t(`DatasetCard.copy`)}}</span>
      <span class="text is-small is-txtGrey" v-if="copySuccessful"><img svg-inline class="copy-icon" src="../../assets/icons/common/copy-success.svg">{{$t(`DatasetCard.copySuccess`)}}</span>
    </div>
</template>

<script>
export default {
  name: 'CopyDropdown',
  props: {
    textToCopy: String,
    isVisible: Boolean
  },
  data () {
    return {
      copySuccessful: false,
      hideDropdownTimeout: null
    };
  },
  watch: {
    isVisible: function (newVal, oldVal) {
      if (newVal === false) {
        this.resetDropdown();
      }
    }
  },
  methods: {
    copyName () {
      var textArea = document.createElement('textarea');
      textArea.style.position = 'fixed';
      textArea.style.opacity = 0;
      textArea.style.height = 0;
      textArea.style.width = 0;
      textArea.value = this.textToCopy;
      document.body.appendChild(textArea);
      textArea.focus();
      textArea.select();
      try {
        this.copySuccessful = document.execCommand('copy');
        if (this.copySuccessful) {
          this.hideDropdownTimeout = setTimeout(() => { this.$emit('hideDropdown'); }, 2000);
        }
      } catch (err) {
        this.copySuccessful = false;
      } finally {
        document.body.removeChild(textArea);
      }
    },
    resetDropdown () {
      this.copySuccessful = false;
      if (this.hideDropdownTimeout) {
        clearTimeout(this.hideDropdownTimeout);
      }
    }
  }
};
</script>

<style lang="scss" scoped>
@import 'new-dashboard/styles/variables';

.dropdown {
  position: relative;
  padding: 12px 16px 8px;
  border: 1px solid $border-color;
  border-radius: 4px;
  background-color: #FFF;

  &::before {
    content: '';
    position: absolute;
    bottom: -8px;
    left: 24px;
    width: 14px;
    height: 14px;
    transform: rotate(45deg);
    border: 1px solid $neutral--200;
    border-top: none;
    border-left: none;
    border-radius: 2px;
    background-color: #FFF;
  }
}

.copy-text {
  text-decoration: underline;
}

.copy-icon {
  margin-right: 8px;
  margin-bottom: -2px;
}
</style>
