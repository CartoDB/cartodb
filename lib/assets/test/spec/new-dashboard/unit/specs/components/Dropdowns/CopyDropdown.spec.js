import { shallowMount } from '@vue/test-utils';
import CopyDropdown from 'new-dashboard/components/Dropdowns/CopyDropdown';

const $t = key => key;

describe('CopyDropdown.vue', () => {
  describe('properties', () => {
    it('should render correct contents', () => {
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });
      expect(copyDropdown).toMatchSnapshot();
    });

    it('should show confirmation message when data is copied', () => {
      document.execCommand = key => true;
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });

      copyDropdown.vm.copyName();

      expect(copyDropdown.vm.copySuccessful).toBe(true);
      expect(copyDropdown).toMatchSnapshot();
    });

    it('should not show confirmation message if data is not copied', () => {
      document.execCommand = key => false;
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });

      copyDropdown.vm.copyName();

      expect(copyDropdown.vm.copySuccessful).toBe(false);
      expect(copyDropdown).toMatchSnapshot();
    });

    it('should handle an error in execCommand', () => {
      document.execCommand = false;
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });

      copyDropdown.vm.copyName();

      expect(copyDropdown.vm.copySuccessful).toBe(false);
      expect(copyDropdown).toMatchSnapshot();
    });

    it('should hide the dropdown if isVisible is set to false', () => {
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });

      copyDropdown.vm.isVisible = false;

      expect(copyDropdown).toMatchSnapshot();
    });

    it('should show copy message again after the method resetDropdown is called', () => {
      document.execCommand = key => true;
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });

      copyDropdown.vm.copyName();
      copyDropdown.vm.resetDropdown();
      expect(copyDropdown.vm.copySuccessful).toBe(false);
      expect(copyDropdown).toMatchSnapshot();
    });
  });

  describe('watchers', () => {
    it('should run the reset method when isVisible prop is changed to false', () => {
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: true
        },
        mocks: {
          $t
        }
      });
      jest.spyOn(copyDropdown.vm, 'resetDropdown');
      copyDropdown.vm.isVisible = false;

      expect(copyDropdown.vm.resetDropdown).toBeCalled();
    });

    it('should not run the reset method when isVisible prop is changed to true', () => {
      const copyDropdown = shallowMount(CopyDropdown, {
        propsData: {
          textToCopy: 'This is a test',
          isVisible: false
        },
        mocks: {
          $t
        }
      });
      jest.spyOn(copyDropdown.vm, 'resetDropdown');
      copyDropdown.vm.isVisible = true;

      expect(copyDropdown.vm.resetDropdown).not.toBeCalled();
    });
  });
});
