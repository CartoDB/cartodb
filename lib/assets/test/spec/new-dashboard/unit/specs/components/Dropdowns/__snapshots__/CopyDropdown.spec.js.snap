// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyDropdown.vue properties should handle an error in execCommand 1`] = `
<div class="dropdown"><span class="text is-small copy-text"><img svg-inline="" src="../../assets/icons/common/copy.svg" class="copy-icon">DatasetCard.copy</span>
  <!---->
</div>
`;

exports[`CopyDropdown.vue properties should hide the dropdown if isVisible is set to false 1`] = ``;

exports[`CopyDropdown.vue properties should not show confirmation message if data is not copied 1`] = `
<div class="dropdown"><span class="text is-small copy-text"><img svg-inline="" src="../../assets/icons/common/copy.svg" class="copy-icon">DatasetCard.copy</span>
  <!---->
</div>
`;

exports[`CopyDropdown.vue properties should render correct contents 1`] = `
<div class="dropdown"><span class="text is-small copy-text"><img svg-inline="" src="../../assets/icons/common/copy.svg" class="copy-icon">DatasetCard.copy</span>
  <!---->
</div>
`;

exports[`CopyDropdown.vue properties should show confirmation message when data is copied 1`] = `
<div class="dropdown">
  <!----> <span class="text is-small is-txtGrey"><img svg-inline="" src="../../assets/icons/common/copy-success.svg" class="copy-icon">DatasetCard.copySuccess</span></div>
`;

exports[`CopyDropdown.vue properties should show copy message again after the method resetDropdown is called 1`] = `
<div class="dropdown"><span class="text is-small copy-text"><img svg-inline="" src="../../assets/icons/common/copy.svg" class="copy-icon">DatasetCard.copy</span>
  <!---->
</div>
`;
